src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1330,23 +1318,33 @@ private void AbortStreams(Exception abortException)
             // we could hold pool lock while trying to grab connection lock in Dispose().
             _pool.InvalidateHttp2Connection(this);
 
+            List<Http2Stream> streamsToAbort = new List<Http2Stream>();
+
             lock (SyncObject)
             {
                 if (NetEventSource.IsEnabled) Trace($"{nameof(abortException)}={abortException}");
 
+                // Set _lastStreamId to int.MaxValue to indicate that we are shutting down
+                // and we must assume all active streams have been processed by the server
+                _lastStreamId = int.MaxValue;
+
                 foreach (KeyValuePair<int, Http2Stream> kvp in _httpStreams)
                 {
                     int streamId = kvp.Key;
                     Debug.Assert(streamId == kvp.Value.StreamId);
 
-                    kvp.Value.OnAbort(abortException);
+                    streamsToAbort.Add(kvp.Value);
```

Comment:
How about under the lock just swapping _httpStreams out for an empty instance (if it's not already empty), and then outside of the loop iterating over the local?  Then we don't need to copy everything over and we allocate less.

Reply:
We can't mutate _httpStreams here. We need to continue to track the aborted streams until they shut down and call us back to remove the stream from the dictionary.